### DISCUSSION TITLE
Allow vite.plugins added to astro.config.js to come before @astrojs plugins

### DISCUSSION THREAD

**Discussion by leoj3n** [withastro/roadmap#120] [Proposal]
The following argument order for

```js
vite.mergeConfig(viteConfig, inlineConfig);
```

in [packages/astro/src/core/create-vite.ts#L97](https://github.com/withastro/astro/blob/137b08efc289111d3fa39ae79dc57f66810978e5/packages/astro/src/core/create-vite.ts#L97) forces plugins added to `vite.plugins` in `astro.config.js` like so:

```js
// @ts-check
import { myPlugin } from '@example/my-plugin'

/** @type {import('astro').AstroUserConfig} */
const config = {
	vite: {
		plugins: [
			myPlugin(),
		],
	},
}

export default config
```

to always come last like that:

```json
    "plugins": [
        {},
        {
            "name": "@astrojs/vite-plugin-astro",
            "enforce": "pre"
        },
        {
            "name": "@astrojs/vite-plugin-markdown",
            "enforce": "pre"
        },
        {
            "name": "@astrojs/vite-plugin-jsx",
            "enforce": "pre"
        },
        {
            "name": "@astrojs/vite-plugin-astro-postprocess"
        },
        {
            "name": "myPlugin",
            "enforce": "pre"
        }
    ],
```

For my use case, I require my plugin to come first.

A solution might allow specifying `beforeAstroPlugins: true` or even "before/after" a specific `@astrojs` plugin, such as:

```
after: '@astrojs/vite-plugin-markdown'
```

This would give the needed flexibility to load a plugin specifically before the `@astrojs` plugins.

How vite orders plugins using `pre` and `post`:

```ts
export function sortUserPlugins(
  plugins: (Plugin | Plugin[])[] | undefined
): [Plugin[], Plugin[], Plugin[]] {
  const prePlugins: Plugin[] = []
  const postPlugins: Plugin[] = []
  const normalPlugins: Plugin[] = []


  if (plugins) {
    plugins.flat().forEach((p) => {
      if (p.enforce === 'pre') prePlugins.push(p)
      else if (p.enforce === 'post') postPlugins.push(p)
      else normalPlugins.push(p)
    })
  }


  return [prePlugins, normalPlugins, postPlugins]
}
```

https://github.com/vitejs/vite/blob/08cf4e106534481fd8eff2e6e4a986270c28f894/packages/vite/src/node/config.ts#L827-L843

**Comment by leoj3n** [comment]
Callback solution...

`packages/astro/src/core/create-vite.ts`:

```js
  // ...
  if (typeof viteConfig.mergedConfigCallback === "function") {
      return viteConfig.mergedConfigCallback(viteConfig);
  }
  
  return viteConfig;
}
```

Usage in `astro.config.js`:

```js
// @ts-check
import replace from '@rollup/plugin-replace';
import {
    myPlugin
} from '@example/my-plugin';

/** @type {import('astro').AstroUserConfig} */
const config = {
    buildOptions: {
        site: 'http://localhost:3000/',
    },
    renderers: ['@astrojs/renderer-react'],
    vite: {
        mergedConfigCallback: (config) => {
            const headPlugins = config.plugins.filter(p => p.beforeAstroPlugins === true);
            const tailPlugins = config.plugins.filter(p => p.beforeAstroPlugins !== true);
            config.plugins = headPlugins.concat(tailPlugins);

            return config;
        },
        plugins: [{
                ...replace({
                    include: ['demo/src/**'],
                    values: {
                        changed: 'replaced',
                    },
                }),
                beforeAstroPlugins: true,
            },
            {
                ...myPlugin(),
                beforeAstroPlugins: true,
            },
        ],
    },
};

export default config;
```

**Comment by natemoo-re** [comment]
Linking to the [Implementation PR](https://github.com/withastro/astro/pull/2611) for posterity.

We do need some answer for this, but I'm not sure if extending the Vite config interface with Astro-specific hooks is the right way to go. 

I'd love to get this RFC back on our radar and get some more eyes on this.

**Reply by FredKSchott** [reply]
Vite already has an answer for this! Check out https://vitejs.dev/guide/using-plugins.html#enforcing-plugin-ordering and the `enforce: 'pre'` config.

**Reply by FredKSchott** [reply]
Oh shoot, I just went back and saw you using `pre` in the config above. 

Unfortunately, this is a problem across the entire Vite (and larger Rollup) ecosystem. You can see us workaround this here in `packages/astro/src/vite-plugin-build-css/index.ts`:

```js
// Remove us from where we are now and place us right before the viteCSSPost plugin
plugins.splice(ourIndex, 1);
plugins.splice(viteCSSPostIndex - 1, 0, ourPlugin);
```

I'd rather wait for Vite's lead on this one, since I know they're aware of it. You may have some luck posting in the Vite issue tracker, as a general ask about controlling plugin order.
